### PR TITLE
[spinel] add CSL Accuracy and CSL Uncertainty properties

### PR DESCRIPTION
**Pull request by edmont**
Allow the host to get CSL Accuracy and CSL Uncertainty from the RCP.

Depends-On: openthread/ot-nrf528xx#313

### CODE REVIEW

File: src/lib/spinel/spinel.c [hunk 1]
```diff
@@ -1272,6 +1272,7 @@ const char *spinel_prop_key_to_cstr(spinel_prop_key_t prop_key)
         {SPINEL_PROP_MAC_CCA_FAILURE_RATE, "MAC_CCA_FAILURE_RATE"},
         {SPINEL_PROP_MAC_MAX_RETRY_NUMBER_DIRECT, "MAC_MAX_RETRY_NUMBER_DIRECT"},
         {SPINEL_PROP_MAC_MAX_RETRY_NUMBER_INDIRECT, "MAC_MAX_RETRY_NUMBER_INDIRECT"},
+        {SPINEL_PROP_MAC_CSL_ACCURACY, "MAC_CSL_ACCURACY"},
```
Comment (bukepo) [inline src/lib/spinel/spinel.c:1275]: Should this be SPINEL_PROP_RCP_CSL_ACCURACY?

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 2]
```diff
@@ -2461,5 +2461,19 @@ otError RadioSpinel<InterfaceType, ProcessContextType>::ConfigureEnhAckProbing(o
 }
 #endif
 
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+template <typename InterfaceType, typename ProcessContextType>
+otError RadioSpinel<InterfaceType, ProcessContextType>::GetCslAccuracy(uint8_t *aAccuracy)
+{
+    otError error = OT_ERROR_NONE;
```
Comment (zhanglongxia) [inline src/lib/spinel/radio_spinel_impl.hpp:2468]: Should we init the `error`?

File: src/core/thread/mle.cpp [hunk 3]
```diff
@@ -1533,7 +1533,7 @@ Error Mle::AppendCslClockAccuracy(Message &aMessage)
 
     cslClockAccuracy.Init();
 
-    cslClockAccuracy.SetCslClockAccuracy(Get<Radio>().GetCslAccuracy());
+    cslClockAccuracy.SetCslClockAccuracy(otPlatRadioGetCslAccuracy(&GetInstance()));
```
Comment (zhanglongxia) [inline src/core/thread/mle.cpp:1536]: The `Get<Radio>().GetCslAccuracy()` calls the `otPlatRadioGetCslAccuracy`. Should we keep the code unchanged?

Comment (Irving-cl) [reply src/core/thread/mle.cpp:1536]: I agree with Zhanglong. We should prefer calling high level functions than platform API.

**Review by Irving-cl — APPROVED** [review]

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 2 above)
Comment (edmont) [reply src/lib/spinel/radio_spinel_impl.hpp]: Not really, removed.

File: src/lib/spinel/spinel.c (hunk 1 above)
Comment (edmont) [reply src/lib/spinel/spinel.c]: Changed.

File: src/core/thread/mle.cpp (hunk 3 above)
Comment (edmont) [reply src/core/thread/mle.cpp]: Changed.

File: src/lib/spinel/radio_spinel.hpp [hunk 4]
```diff
@@ -701,6 +701,22 @@ template <typename InterfaceType, typename ProcessContextType> class RadioSpinel
...
+     * @note Platforms may optimize this value based on operational conditions (i.e.: temperature).
+     *
+     * @param[out]   aAccuracy  The current CSL rx/tx scheduling drift, in units of ± ppm
+     *
+     * @retval  OT_ERROR_INVALID_ARGS     @p aAccuracy is nullptr.
+     * @retval  OT_ERROR_NONE             Successfully got CSL Accuracy.
+     * @retval  OT_ERROR_FAILED           Other platform specific errors.
+     *
+     */
+    otError GetCslAccuracy(uint8_t *aAccuracy);
```
Comment (bukepo) [inline src/lib/spinel/radio_spinel.hpp:717]:
```suggestion
    otError GetCslAccuracy(uint8_t &aAccuracy);
```

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 5]
```diff
@@ -2461,5 +2461,19 @@ otError RadioSpinel<InterfaceType, ProcessContextType>::ConfigureEnhAckProbing(o
 }
 #endif
 
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+template <typename InterfaceType, typename ProcessContextType>
+otError RadioSpinel<InterfaceType, ProcessContextType>::GetCslAccuracy(uint8_t *aAccuracy)
```
Comment (bukepo) [inline src/lib/spinel/radio_spinel_impl.hpp:2466]:
```suggestion
otError RadioSpinel<InterfaceType, ProcessContextType>::GetCslAccuracy(uint8_t &aAccuracy)
```

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/lib/spinel/radio_spinel.hpp (hunk 4 above)
Comment (edmont) [reply src/lib/spinel/radio_spinel.hpp]: Done.

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 5 above)
Comment (edmont) [reply src/lib/spinel/radio_spinel_impl.hpp]: Done.

**Review by bukepo — APPROVED** [review]

**Comment by edmont** [comment]
Should `SPINEL_RCP_API_VERSION` and/or `SPINEL_MIN_HOST_SUPPORTED_RCP_API_VERSION` be incremented?

File: src/lib/spinel/spinel.h [hunk 6]
```diff
@@ -4731,6 +4731,15 @@ enum
...
     SPINEL_PROP_RCP_ENH_ACK_PROBING = SPINEL_PROP_RCP_EXT__BEGIN + 3,
 
+    /// CSL Accuracy
+    /** Format: `c`
+     * Required capability: `SPINEL_CAP_NET_THREAD_1_2`
+     *
+     * The current CSL rx/tx scheduling drift, in units of ± ppm.
+     *
+     */
+    SPINEL_PROP_RCP_CSL_ACCURACY = SPINEL_PROP_RCP_EXT__BEGIN + 4,
```
[the remark this answers is not in the document]
Comment (edmont) [reply src/lib/spinel/spinel.h:4741]: Yes, according to the _Spinel definition compatibility guideline_ I believe only the fist one should be increased.

**Comment by edmont** [comment]
Adding similar changes for the CSL Uncertainty in the same PR so that `SPINEL_RCP_API_VERSION` doesn't have to be updated so often.

BTW, it seems some of the expect tests are failing on main.

**Comment by edmont** [comment]
How is it possible that the tests are passing for me locally but they fail on the CI?

File: src/core/config/platform.h [hunk 7]
```diff
@@ -166,14 +166,4 @@
...
 #endif
 
-/**
- * @def OPENTHREAD_CONFIG_PLATFORM_CSL_UNCERT
- *
- * The Uncertainty of the scheduling CSL of transmission by the parent, in ±10 us units.
- *
- */
-#ifndef OPENTHREAD_CONFIG_PLATFORM_CSL_UNCERT
-#define OPENTHREAD_CONFIG_PLATFORM_CSL_UNCERT UINT8_MAX
```
Comment (abtink) [inline src/core/config/platform.h:176]: Why is this config removed?

If we want to remove it we need to make sure if any project/build that used this noticed this removal.
For this, we add a check in `openthread-core-config-check` to give error if it is being used.

File: src/lib/spinel/spinel.h [hunk 8]
```diff
@@ -4731,6 +4731,24 @@ enum
      */
     SPINEL_PROP_RCP_ENH_ACK_PROBING = SPINEL_PROP_RCP_EXT__BEGIN + 3,
 
+    /// CSL Accuracy
+    /** Format: `c`
```
Comment (abtink) [inline src/lib/spinel/spinel.h:4735]: Should it be capital `"C"`? (`"C"` is uint8 and `"c"` is int8).

File: src/lib/spinel/spinel.h [hunk 9]
```diff
@@ -4731,6 +4731,24 @@ enum
...
+     * Required capability: `SPINEL_CAP_NET_THREAD_1_2`
+     *
+     * The current CSL rx/tx scheduling drift, in units of ± ppm.
+     *
+     */
+    SPINEL_PROP_RCP_CSL_ACCURACY = SPINEL_PROP_RCP_EXT__BEGIN + 4,
+
+    /// CSL Uncertainty
+    /** Format: `c`
```
Comment (abtink) [inline src/lib/spinel/spinel.h:4744]: Same here?

**Review by abtink — APPROVED** [review]
Thanks @edmont .Looks great.
Few smaller suggestions/comments below.

File: src/core/config/platform.h (hunk 7 above)
Comment (edmont) [reply src/core/config/platform.h:176]: I'm fine with leaving it, but found difficult to make use of this config for Posix platform. Anyway platforms should implement the weak `otPlatRadioGetCslClockUncertainty` and eventually this function could return some variable result depending on the device conditions.

File: src/lib/spinel/spinel.h (hunk 8 above)
Comment (edmont) [reply src/lib/spinel/spinel.h:4735]: Right, should be `uint8_t`.

**Comment by jwhui** [comment]
> @jwhui can this be merged? The following tests are passing for me locally on the PR's branch:
> 
> ```
> PASS 1 FAIL 0 tests/scripts/thread-cert/border_router/test_trel_connectivity.py
> PASS 1 FAIL 0 tests/scripts/thread-cert/v1_2_LowPower_5_3_01_SSEDAttachment.py
> PASS tests/scripts/expect/v1_2-sed_mac_send.exp
> ```

@edmont , I am not able to get the GitHub Action checks to pass even after a few re-runs.

**Comment by edmont** [comment]
I was missing `OT_NODE_TYPE=rcp` when running test locally. Hopefully issues should be fixed now.

**Review by jwhui — APPROVED** [review]
Thanks! 👍🏼

**Pull request merged**